File: backend/models/frpc.go
package models

type GitHubRelease struct {
	TagName     string               `json:"tag_name"`
	Name        string               `json:"name"`
	HTMLURL     string               `json:"html_url"`
	PublishedAt string               `json:"published_at"`
	Assets      []GitHubReleaseAsset `json:"assets"`
}

type GitHubReleaseAsset struct {
	Name               string `json:"name"`
	BrowserDownloadURL string `json:"browser_download_url"`
	Size               int64  `json:"size"`
	ContentType        string `json:"content_type"`
	Digest             string `json:"digest"`
}

type FrpcReleaseAsset struct {
	Name          string `json:"name"`
	DownloadURL   string `json:"download_url"`
	ContentType   string `json:"content_type"`
	Size          int64  `json:"size"`
	Digest        string `json:"digest"`
	SHA256        string `json:"sha256"`
	OS            string `json:"os"`
	Arch          string `json:"arch"`
	ArchiveFormat string `json:"archive_format"`
}

type FrpcReleaseInfo struct {
	TagName     string           `json:"tag_name"`
	Name        string           `json:"name"`
	HTMLURL     string           `json:"html_url"`
	PublishedAt string           `json:"published_at"`
	Asset       FrpcReleaseAsset `json:"asset"`
}

type FrpcInstalledInfo struct {
	Version      string `json:"version"`
	AssetName    string `json:"asset_name"`
	SHA256       string `json:"sha256"`
	InstalledAt  string `json:"installed_at"`
	BinaryPath   string `json:"binary_path"`
	BinaryExists bool   `json:"binary_exists"`
}

type FrpcPaths struct {
	UserDataDir  string `json:"userdata_dir"`
	FrpcDir      string `json:"frpc_dir"`
	BinDir       string `json:"bin_dir"`
	BinaryPath   string `json:"binary_path"`
	DownloadDir  string `json:"download_dir"`
	StatePath    string `json:"state_path"`
	SettingsPath string `json:"settings_path"`
}

type FrpcStatus struct {
	GOOS            string             `json:"goos"`
	GOARCH          string             `json:"goarch"`
	Paths           FrpcPaths          `json:"paths"`
	GitHubMirrorURL string             `json:"github_mirror_url"`
	Installed       *FrpcInstalledInfo `json:"installed,omitempty"`
	Latest          *FrpcReleaseInfo   `json:"latest,omitempty"`
	UpdateAvailable bool               `json:"update_available"`
	LatestError     string             `json:"latest_error,omitempty"`
}

type FrpcInstallResult struct {
	Release FrpcReleaseInfo `json:"release"`
	Status  FrpcStatus      `json:"status"`
}

File: frontend/src/services/frpc.ts
type FrpcServiceBinding = {
  GetFrpcStatus: () => Promise<any>;
  GetGitHubMirrorURL: () => Promise<string>;
  InstallOrUpdateFrpc: () => Promise<any>;
  RemoveFrpc: () => Promise<void>;
  SetGitHubMirrorURL: (url: string) => Promise<void>;
};

function getFrpcServiceBinding(): FrpcServiceBinding {
  const svc = (window as any).go?.services?.FrpcService;
  if (!svc) {
    throw new Error("FrpcService 未绑定，请重启应用。");
  }
  return svc as FrpcServiceBinding;
}

function parseError(error: unknown): Error {
  if (error instanceof Error) {
    return error;
  }
  if (typeof error === "string") {
    return new Error(error);
  }
  if (typeof error === "object" && error !== null && "message" in error) {
    const message = (error as { message?: unknown }).message;
    if (typeof message === "string") {
      return new Error(message);
    }
  }
  return new Error("请求失败");
}

export interface FrpcPaths {
  userdata_dir: string;
  frpc_dir: string;
  bin_dir: string;
  binary_path: string;
  download_dir: string;
  state_path: string;
  settings_path: string;
}

export interface FrpcInstalledInfo {
  version: string;
  asset_name: string;
  sha256: string;
  installed_at: string;
  binary_path: string;
  binary_exists: boolean;
}

export interface FrpcReleaseAsset {
  name: string;
  download_url: string;
  content_type: string;
  size: number;
  digest: string;
  sha256: string;
  os: string;
  arch: string;
  archive_format: string;
}

export interface FrpcReleaseInfo {
  tag_name: string;
  name: string;
  html_url: string;
  published_at: string;
  asset: FrpcReleaseAsset;
}

export interface FrpcStatus {
  goos: string;
  goarch: string;
  paths: FrpcPaths;
  github_mirror_url: string;
  installed?: FrpcInstalledInfo;
  latest?: FrpcReleaseInfo;
  update_available: boolean;
  latest_error?: string;
}

export interface FrpcInstallResult {
  release: FrpcReleaseInfo;
  status: FrpcStatus;
}

export async function getFrpcStatus(): Promise<FrpcStatus> {
  try {
    const svc = getFrpcServiceBinding();
    return (await svc.GetFrpcStatus()) as FrpcStatus;
  } catch (error) {
    throw parseError(error);
  }
}

export async function getGitHubMirrorURL(): Promise<string> {
  try {
    const svc = getFrpcServiceBinding();
    return (await svc.GetGitHubMirrorURL()) as string;
  } catch (error) {
    throw parseError(error);
  }
}

export async function installOrUpdateFrpc(): Promise<FrpcInstallResult> {
  try {
    const svc = getFrpcServiceBinding();
    return (await svc.InstallOrUpdateFrpc()) as FrpcInstallResult;
  } catch (error) {
    throw parseError(error);
  }
}

export async function removeFrpc(): Promise<void> {
  try {
    const svc = getFrpcServiceBinding();
    await svc.RemoveFrpc();
  } catch (error) {
    throw parseError(error);
  }
}

export async function setGitHubMirrorURL(url: string): Promise<void> {
  try {
    const svc = getFrpcServiceBinding();
    await svc.SetGitHubMirrorURL(url);
  } catch (error) {
    throw parseError(error);
  }
}
